File: app/src/main/java/dev/carloszuil/herojourney/adapter/HabitExpandableAdapter.java
package dev.carloszuil.herojourney.adapter;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CheckBox;
import android.widget.FrameLayout;
import android.widget.ImageSwitcher;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.BiConsumer;
import java.util.function.Consumer;

import dev.carloszuil.herojourney.R;
import dev.carloszuil.herojourney.data.local.entities.Habit;

public class HabitExpandableAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private final Consumer<Habit> onHabitToggled;
    private final BiConsumer<String, Boolean> onSectionToggled;
    private final Consumer<Habit> onHabitClicked;

    private final List<HabitListItem> items = new ArrayList<>();
    private final Map<String, Boolean> sectionExpandedState = new HashMap<>();
    private List<Habit> allHabits = new ArrayList<>();

    public HabitExpandableAdapter(
            Consumer<Habit> onHabitToggled,
            BiConsumer<String, Boolean> onSectionToggled,
            Consumer<Habit> onHabitClicked
    ) {
        this.onHabitToggled = onHabitToggled;
        this.onSectionToggled = onSectionToggled;
        this.onHabitClicked = onHabitClicked;
    }

    public void submitHabits(@NonNull List<Habit> habits,
                             boolean pendientesExpanded,
                             boolean completadasExpanded) {
        this.allHabits = new ArrayList<>(habits);
        sectionExpandedState.put("📌 Pendientes", pendientesExpanded);
        sectionExpandedState.put("✅ Completadas", completadasExpanded);
        rebuildItems();
    }

    private void rebuildItems() {
        List<Habit> pendientes = new ArrayList<>();
        List<Habit> completadas = new ArrayList<>();
        for (Habit h : allHabits) {
            if (h.isFinished()) completadas.add(h);
            else pendientes.add(h);
        }

        items.clear();
        boolean pendExp = sectionExpandedState.getOrDefault("📌 Pendientes", true);
        items.add(new HabitListItem.SectionHeader("📌 Pendientes", pendExp));
        if (pendExp) {
            for (Habit h : pendientes) items.add(new HabitListItem.HabitItem(h));
        }
        boolean compExp = sectionExpandedState.getOrDefault("✅ Completadas", false);
        items.add(new HabitListItem.SectionHeader("✅ Completadas", compExp));
        if (compExp) {
            for (Habit h : completadas) items.add(new HabitListItem.HabitItem(h));
        }
        notifyDataSetChanged();
    }

    @Override
    public int getItemViewType(int position) {
        return (items.get(position) instanceof HabitListItem.SectionHeader) ? 0 : 1;
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View v = LayoutInflater.from(parent.getContext())
                .inflate(viewType == 0 ? R.layout.item_section_header : R.layout.item_habit,
                        parent, false);
        return viewType == 0 ? new SectionViewHolder(v) : new HabitViewHolder(v);
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        HabitListItem item = items.get(position);
        if (holder instanceof SectionViewHolder) {
            ((SectionViewHolder) holder).bind((HabitListItem.SectionHeader) item);
        } else {
            ((HabitViewHolder) holder).bind(((HabitListItem.HabitItem) item).getHabit());
        }
    }

    @Override
    public int getItemCount() {
        return items.size();
    }

    class SectionViewHolder extends RecyclerView.ViewHolder {
        private final TextView title;
        private final ImageSwitcher switcher;

        SectionViewHolder(@NonNull View view) {
            super(view);
            title = view.findViewById(R.id.sectionTitle);
            switcher = view.findViewById(R.id.expandSwitcher);
            switcher.setFactory(() -> {
                ImageView iv = new ImageView(view.getContext());
                iv.setLayoutParams(new FrameLayout.LayoutParams(
                        FrameLayout.LayoutParams.WRAP_CONTENT,
                        FrameLayout.LayoutParams.WRAP_CONTENT));
                iv.setScaleType(ImageView.ScaleType.CENTER_INSIDE);
                return iv;
            });
        }

        void bind(HabitListItem.SectionHeader header) {
            title.setText(header.getTitle());
            int icon = header.isExpanded() ? R.drawable.ic_expand_less : R.drawable.ic_expand_more;
            switcher.setImageResource(icon);
            itemView.setOnClickListener(v -> {
                boolean newState = !header.isExpanded();
                header.setExpanded(newState);
                sectionExpandedState.put(header.getTitle(), newState);
                rebuildItems();
                onSectionToggled.accept(header.getTitle(), newState);
            });
        }
    }

    class HabitViewHolder extends RecyclerView.ViewHolder {
        private final CheckBox check;
        private final TextView text;

        HabitViewHolder(@NonNull View view) {
            super(view);
            check = view.findViewById(R.id.checkboxTarea);
            text = view.findViewById(R.id.textoTarea);
        }

        void bind(Habit habit) {
            text.setText(habit.getName());
            check.setOnCheckedChangeListener(null);
            check.setChecked(habit.isFinished());
            check.setOnCheckedChangeListener((btn, isChecked) -> {
                habit.setFinished(isChecked);
                rebuildItems();
                onHabitToggled.accept(habit); // notifica el hábito cambiado
            });
            itemView.setOnClickListener(v -> onHabitClicked.accept(habit));
        }
    }
}

File: app/src/main/java/dev/carloszuil/herojourney/adapter/HabitExpandableAdapter.kt
package dev.carloszuil.herojourney.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.CheckBox
import android.widget.FrameLayout
import android.widget.ImageSwitcher
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import dev.carloszuil.herojourney.R
import dev.carloszuil.herojourney.model.Habit

class HabitExpandableAdapter(
    private val onHabitToggled: () -> Unit,
    private val onSectionToggled: (String, Boolean) -> Unit,
    private val onHabitClicked: (Habit) -> Unit
) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    private val items = mutableListOf<HabitListItem>()
    private val sectionExpandedState = mutableMapOf<String, Boolean>()
    private var allHabits = listOf<Habit>()

    fun submitHabits(habits: List<Habit>, pendExpanded: Boolean, compExpanded: Boolean) {
        allHabits = habits
        sectionExpandedState["📌 Pendientes"] = pendExpanded
        sectionExpandedState["✅ Completadas"] = compExpanded
        rebuildItems()
    }

    private fun rebuildItems() {
        val pendientes = allHabits.filter { !it.completada }
        val completadas = allHabits.filter { it.completada }

        items.clear()
        val pendExp = sectionExpandedState.getOrDefault("📌 Pendientes", true)
        items.add(HabitListItem.SectionHeader("📌 Pendientes", pendExp))
        if (pendExp) items += pendientes.map { HabitListItem.HabitItem(it) }

        val compExp = sectionExpandedState.getOrDefault("✅ Completadas", false)
        items.add(HabitListItem.SectionHeader("✅ Completadas", compExp))
        if (compExp) items += completadas.map { HabitListItem.HabitItem(it) }

        notifyDataSetChanged()
    }

    override fun getItemViewType(position: Int): Int = when(items[position]) {
        is HabitListItem.SectionHeader -> 0
        is HabitListItem.HabitItem -> 1
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return if (viewType == 0) {
            val view = LayoutInflater.from(parent.context)
                .inflate(R.layout.item_section_header, parent, false)
            SectionViewHolder(view)
        } else {
            val view = LayoutInflater.from(parent.context)
                .inflate(R.layout.item_habit, parent, false)
            HabitViewHolder(view)
        }
    }

    override fun getItemCount(): Int = items.size

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        when (val item = items[position]) {
            is HabitListItem.SectionHeader -> (holder as SectionViewHolder).bind(item)
            is HabitListItem.HabitItem -> (holder as HabitViewHolder).bind(item.habit)
        }
    }

    inner class SectionViewHolder(view: View): RecyclerView.ViewHolder(view) {
        private val title: TextView = view.findViewById(R.id.sectionTitle)
        private val switcher: ImageSwitcher = view.findViewById(R.id.expandSwitcher)

        init {
            switcher.setFactory {
                ImageView(view.context).apply {
                    layoutParams = FrameLayout.LayoutParams(
                        ViewGroup.LayoutParams.WRAP_CONTENT,
                        ViewGroup.LayoutParams.WRAP_CONTENT
                    )
                    scaleType = ImageView.ScaleType.CENTER_INSIDE
                }
            }
        }

        fun bind(header: HabitListItem.SectionHeader) {
            title.text = header.title
            val icon = if (header.isExpanded) R.drawable.ic_expand_less else R.drawable.ic_expand_more
            switcher.setImageResource(icon)
            itemView.setOnClickListener {
                header.isExpanded = !header.isExpanded
                sectionExpandedState[header.title] = header.isExpanded
                rebuildItems()
                onSectionToggled(header.title, header.isExpanded)
            }
        }
    }

    inner class HabitViewHolder(view: View): RecyclerView.ViewHolder(view) {
        private val check: CheckBox = view.findViewById(R.id.checkboxTarea)
        private val text: TextView = view.findViewById(R.id.textoTarea)

        fun bind(habit: Habit) {
            text.text = habit.nombre
            check.setOnCheckedChangeListener(null)
            check.isChecked = habit.completada
            check.setOnCheckedChangeListener { _, isChecked ->
                habit.completada = isChecked
                rebuildItems()
                onHabitToggled()
            }
            itemView.setOnClickListener {
                onHabitClicked(habit)
            }
        }
    }
}
